src/components/ProjectPinButton/ProjectPinButton.tsx

```diff
@@ -0,0 +1,58 @@
+import { useMutation } from '@apollo/client';
+import { IconButton, makeStyles } from '@material-ui/core';
+import { FC, useCallback, useState } from 'react';
+import * as React from 'react';
+import { GQLOperations } from '../../api';
+import { PushPinIconFilled, PushPinIconOutlined } from '../Icons';
+import { TogglePinnedDocument } from './ProjectPinButton.generated';
+
+const useStyles = makeStyles(() => {
+  return {
+    pinIcon: {
+      padding: 10,
+    },
+  };
+});
+
+export interface ProjectPinButtonProps {
+  projectId?: string;
+  pinned?: boolean;
+  className?: string;
+}
+
+export const ProjectPinButton: FC<ProjectPinButtonProps> = ({
```

Comment:
Ok several things here. We can decouple this button from projects, and use Apollo to store the state.
```gql
fragment TogglePin on Pinnable {
  id
  pinned
}
```
I missed `id` here in the API, but it should be added.

This fragment allows us two things.
1) we can add it to our other fragments
```gql
fragment ProjectListItem on Project {
  ...TogglePin
}
fragment ProjectOverview on Project {
  ...TogglePin
}
```
2) we can reference this fragment in our props

```tsx
export interface TogglePinButtonProps {
  object: TogglePinFragment;
  ...
}
```
```tsx
props.object.pinned ? <Filled /> : <Outline />
```

3) With this fragment we can update the cache directly instead of having to do refetch queries.
```tsx
const [togglePinned] = useMutation(TogglePinnedDocument, {
  variables: {
    id: props.object.id,
  },
  update: (cache, result) => {
    if (!result.data) {
      return;
    }
    cache.modify({
      id: cache.identify(props.object),
      fields: {
        pinned: () => result.data.togglePinned,
      },
    });
  },
});
```
(I haven't tested exactly so it may need some tweaks)

Then we can just pass the function right to the button handler
```tsx
<IconButton onClick={togglePinned}
```

We don't need to manage state ourselves because Apollo cache will handle that.

Reply:
Oh we probably need to add the project to the pinned list if we can find it. I can help with that later. I'm working on the list cache right now.

Reply:
```
cache.identify(props.object)
```
It returns `undefined` value because we don't cache `TogglePinFragment ` yet.
FYI, I passed `project` in this component, and `cache.identify(props.project)` will return correct id and it updates the pinned field correctly.

Reply:
Aha, I can see what you meant. NVM